### PR TITLE
Upgrade an entire cluster from 5 to 6 for a single heap table.

### PR DESCRIPTION
**Pull request by berlin-ab**
Enable upgrade of a 5x cluster to a 6x cluster.

- changes the ordering of update query on pg_class to ensure
  datfrozenxids are set before vacuum

- adds integration test that makes use of a local installation of both
  versions to perform an upgrade from one cluster to another

- ensures that pg_upgrade connections to both 5x and 6x clusters use
  the correct parameters based on dispatcher vs. segment.

Notes: 

- this test is not exhaustive and can certainly be improved. This, I think, is a nice starting point  
   and sets up follow-on parallelizable work.
- automation can be improved to help set up the development environment necessary to run the 
   suite
- a continuous integration environment still needs to be created for this suite

### CODE REVIEW

**Comment by berlin-ab** [comment]
For debugging purposes, I've added two scripts which are responsible for setting up the correct environment to run gpstart/stop:

`scripts/gpdb5-cluster [start|stop]`

and 

`scripts/gpdb6-cluster [start|stop]`

**Comment by berlin-ab** [comment]
The 9.4 merge FIXME could be removed because the change that we were waiting to be merged has been introduced.

File: contrib/pg_upgrade/pg_upgrade.c [hunk 1]
```diff
@@ -1004,14 +993,28 @@ set_frozenxids(bool minmxid_only)
...
+		/*
+		 * Instead of assuming template0 will be frozen by initdb, its worth
+		 * making sure we freeze it here before updating the relfrozenxid
+		 * directly for the tables in pg_class and datfrozenxid for the
+		 * database in pg_database. Its fast and safe worth than assuming for
+		 * template0.
+		 */
+		if (!minmxid_only && strcmp(datallowconn, "f") == 0)
+		{
+			PQclear(executeQueryOrDie(conn, "VACUUM FREEZE"));
```
Comment (asimrp) [inline contrib/pg_upgrade/pg_upgrade.c]: Our aim is to have `vacuum freeze` as the last command run before upgrade.  If that's correct, should another `vacuum freeze pg_database` be added after the update command below on line 1023?  The transaction IDs from old cluster don't mean anything, rather, they may cause confusion, in the upgraded cluster.

File: contrib/pg_upgrade/test/integration/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        make -C /some/path/to/gpdb5/source/gpAux/gpdemo && \
+        gpstop -a; # ensure cluster is stopped
+
+    source configuration/gpdb6-env.sh && \ 
+        source ./gpdb6/greenplum_path.sh && \
+        make -C /some/path/to/gpdb6/source/gpAux/gpdemo && \
+        gpstop -a; # ensure cluster is stopped
+        
+    # make a backup of the original data directories
+    cp -r gpdb5-data gpdb5-data-copy
```
Comment (asimrp) [inline contrib/pg_upgrade/test/integration/README.md]: What are `gpdb[56]-data`? Are these symlinks to data directories of respective demo clusters?

File: contrib/pg_upgrade/test/integration/utilities/gpdb6-cluster.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#include "stdlib.h"
+
+
+#include "gpdb6-cluster.h"
+
+static void copy_file_from_backup_to_datadir(char *filename, char *segment_path)
+{
+	char *buffer[2000];
```
Comment (danielgustafsson) [inline contrib/pg_upgrade/test/integration/utilities/gpdb6-cluster.c]: I assume this was supposed to be `char buffer[2000]`?

File: contrib/pg_upgrade/test/integration/utilities/query-helpers.c [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * implements:
+ */
+#include "query-helpers.h"
+#include "libpq-fe.h"
+
+
+PGconn *
+connectTo(int port)
+{
+	char *buffer[1000];
```
Comment (danielgustafsson) [inline contrib/pg_upgrade/test/integration/utilities/query-helpers.c]: `char buffer[1000]`?

File: contrib/pg_upgrade/test/integration/utilities/query-helpers.c [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	sprintf(buffer, "dbname=postgres user=adamberlin port=%d", port);
```
Comment (danielgustafsson) [inline contrib/pg_upgrade/test/integration/utilities/query-helpers.c]: Hardcoded user?

File: contrib/pg_upgrade/pg_upgrade.c [hunk 6]
```diff
@@ -1004,14 +993,28 @@ set_frozenxids(bool minmxid_only)
 									  "AND NOT relfrozenxid = 0) "
 									  "OR (relkind IN ('t', 'o', 'b', 'M'))",
 									  old_cluster.controldata.chkpnt_nxtxid));
+		}
 
 		/* set pg_class.relminmxid */
 		PQclear(executeQueryOrDie(conn,
-								  "UPDATE	pg_catalog.pg_class "
-								  "SET	relminmxid = '%u' "
-		/* only heap, materialized view, and TOAST are vacuumed */
```
Comment (danielgustafsson) [inline contrib/pg_upgrade/pg_upgrade.c]: This indentation is from upstream, changing will just cause pointless merge conflicts.

File: contrib/pg_upgrade/pg_upgrade.c [hunk 7]
```diff
@@ -972,19 +972,8 @@ set_frozenxids(bool minmxid_only)
...
-		 * database in pg_database. Its fast and safe worth than assuming for
-		 * template0.
-		 */
-		if (!minmxid_only && strcmp(datallowconn, "f") == 0)
-		{
-			PQclear(executeQueryOrDie(conn, "VACUUM FREEZE"));
-		}
-
 		if (!minmxid_only)
+		{
```
Comment (danielgustafsson) [inline contrib/pg_upgrade/pg_upgrade.c]: This deviates from upstream and might just cause merge conflicts.

File: contrib/pg_upgrade/pg_upgrade.c (hunk 6 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/pg_upgrade.c]: will fix.

File: contrib/pg_upgrade/pg_upgrade.c (hunk 7 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/pg_upgrade.c]: Will remove.

File: contrib/pg_upgrade/test/integration/README.md (hunk 2 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/README.md]: They are fresh installations of demo clusters created by the above commands. The `configuration/gpdbN-env.sh` files specify the DATADIRS path to have make cluster create the cluster at those paths.

File: contrib/pg_upgrade/test/integration/greenplum_five_to_greenplum_six_upgrade_test.c [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+	PQfinish(connection);
+	stopGpdbSixCluster();
+}
+
+
+int
+main(int argc, char *argv[])
+{
+	cmockery_parse_arguments(argc, argv);
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/greenplum_five_to_greenplum_six_upgrade_test.c]: Nothing. I was calling this as an assumed part of the plumbing for cmockery.

File: contrib/pg_upgrade/test/integration/utilities/query-helpers.c (hunk 5 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/utilities/query-helpers.c]: Oops. Will fix.

File: contrib/pg_upgrade/test/integration/utilities/gpdb6-cluster.c (hunk 3 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/utilities/gpdb6-cluster.c]: Yep. Silly mistake - but the tests still passed (accidentally)

**Comment by berlin-ab** [comment]
> How is this test integrated with the top level `installcheck-world` target?

It's not currently, and will not be as a part of this PR. We should add it soon.

File: contrib/pg_upgrade/test/integration/utilities/upgrade-helpers.c [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	       "--mode=dispatcher "
+	       "--old-bindir=./gpdb5/bin "
+	       "--new-bindir=./gpdb6/bin "
+	       "--old-datadir=./gpdb5-data/qddir/demoDataDir-1 "
+	       "--new-datadir=./gpdb6-data/qddir/demoDataDir-1 "
+	);
+}
+
+void
+upgradeContentId0()
```
Comment (soumyadeep2007) [inline contrib/pg_upgrade/test/integration/utilities/upgrade-helpers.c]: Just have `upgradeContentIdN`?

Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/utilities/upgrade-helpers.c]: Good suggestion. Applied.

File: contrib/pg_upgrade/pg_upgrade.c (hunk 7 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/pg_upgrade.c]: Removed.

File: contrib/pg_upgrade/test/integration/greenplum_five_to_greenplum_six_upgrade_test.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/greenplum_five_to_greenplum_six_upgrade_test.c]: cmockery allows for some command line arguments to be passed to the test executable. Seems ok to leave this as a part of the setup.

File: contrib/pg_upgrade/pg_upgrade.c (hunk 1 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/pg_upgrade.c]: Good point. I'll move it.

**Comment by berlin-ab** [comment]
Thanks for the feedback. I have addressed any of the blocking issues. Merging.

**Pull request merged**